python/TestHarness/schedulers/Scheduler.py

```diff
@@ -0,0 +1,539 @@
+
+    # Delete and return downstream nodes based on tester status and TestHarness options
+    # TODO: change the method name, as it doesn't always delete downstream tests
+    def deleteDownstreamTests(self, job_container):
+        with self.thread_lock:
+            failed_job_containers = set([])
+            tester = job_container.getTester()
+            job_dag = job_container.getDAG()
+            if (tester.isFinished() and not tester.didPass() and not tester.isSilent() and not self.skipPrereqs()) \
+                or (self.options.dry_run and not tester.isSilent()):
+
+                # Ask the DAG to delete and return the downstream jobs associated with this job
+                failed_job_containers.update(job_dag.delete_downstreams(job_container))
+
+            for failed_job in failed_job_containers:
+                tester = failed_job.getTester()
+                tester.setStatus('skipped dependency', tester.bucket_skip)
+
+        return failed_job_containers
+
+    # Iterate over supplied tester data container dict and add edge nodes to the DAG (create dependencies)
+    # Note: This method ignores all errors
+    def createEdgeNodes(self, job_container_dict):
+        for tester_name, job_container in job_container_dict.iteritems():
+            tester = job_container.getTester()
+            job_dag = job_container.getDAG()
+
+            # Attempt to add this node to the DAG (prereq iteration may have already added this node)
+            job_dag.add_node_if_not_exists(job_container)
+
+            # Create the edge dependency DAG. Handle any Cyclic errors as passes (but we set the actual test as a failure).
+            for prereq in tester.getPrereqs():
+
+                # we do not care about missing prereqs
+                if prereq not in job_container_dict.keys():
+                    continue
+                else:
+                    # Attempt to add the prereq node if not already exists (main loop might have already added this node)
+                    job_dag.add_node_if_not_exists(job_container_dict[prereq])
+
+                    # Attempt to create an edge node dependency
+                    try:
+                        job_dag.add_edge(job_container_dict[prereq], job_container)
+
+                    # we do not care about cyclic errors
+                    except dag.DAGValidationError:
+                        pass
+
+        return job_container_dict
+
+    # Method to return failing testers which can only be discovered when scrutinized as a group (race conditions for example)
+    def checkGroupFailures(self, job_container_dict):
+        temp_dag = dag.DAG()
+        failed_or_skipped_testers = set([])
+
+        # Create DAG independent nodes
+        for tester_name, job_container in job_container_dict.iteritems():
+            tester = job_container.getTester()
+
+            # If this tester is not runnable, continue to the next tester
+            if not tester.getRunnable(self.options):
```

Comment:
Another "if not foo then continue else do stuff". You can always switch it around and just do "if foo then do stuff"